text/0000-dockerfiles.md

```diff
@@ -0,0 +1,113 @@
+# Meta
+[meta]: #meta
+- Name: Support Dockerfiles
+- Start Date: 2021-06-30
+- Author(s): sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
+- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)
+
+# Summary
+[summary]: #summary
+
+This RFC introduces functionality for customizing base images, as an alternative to stackpacks.
+
+# Motivation
+[motivation]: #motivation
+
+Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with multi-purpose build-time and runtime Dockerfiles.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, buildah, BuildKit, the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
+
+### App-specified Dockerfiles
+
+A buildpack app may have a build.Dockerfile and/or run.Dockerfile in its app directory. A run.Dockerfile is applied to the selected runtime base image after the detection phase. A build.Dockerfile is applied to the build-time base image before the detection phase.
+
+Both Dockerfiles must accept `base_image` and `build_id` args. The `base_image` arg allows the lifecycle to specify the original base image. The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`.  
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of app-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before `run.Dockerfile` is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
+
+### Platform-specified Dockerfiles
+
+The same Dockerfiles may be used to create new base images or modify existing base images outside of the app build process. For both app-specified and base-image-modifying Dockerfiles, any specified labels override existing values.
+
+Dockerfiles that are used to create a base image must create a `/cnb/image/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after any run.Dockerfile or build.Dockerfile is applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.
```

Comment:
Pardon my ignorance in the realm of bill-of-materials, but why not just use a flat file (i.e. one package per line) or simple JSON format? Writing and parsing would seem to be easier that way, reducing the difficulty of creating `genpkgs`. 

What's the driver for CycloneDX? Is there a compatibility concern with external tools? Is the more complicated format required to represent the information? Is there a general tool that one could just drop in here as `genpkgs` and make it work? 

My thought is that having to write a tool for `genpkgs` and use the CycloneDX format is a hurdle for someone wanting to make their own base images.

Reply:
> My thought is that having to write a tool for `genpkgs` and use the CycloneDX format is a hurdle for someone wanting to make their own base images.

I fully agree. Otherwise, buildpack should provide a tool, documentation, ... to generate it automatically using list of packages generated as yaml text file

Reply:
A few points:
1. It's important that genpkgs be an executable, so that the SBoM can be updated automatically when the base image is extended (including with app-specified Dockerfiles)
2. It's important that the format be CycloneDX, given https://github.com/buildpacks/rfcs/pull/166. The goal is to reliably produce an SBoM in an industry-standard and security-focused format.
3. CycloneDX appears to have robust tooling for both security scanning and SBoM generation, e.g., https://github.com/CycloneDX/cyclonedx-maven-plugin
4. I agree that the project could provide genpkgs for common linux distros -- `pack create-stack` could even have a flag to add it automatically

(Will add these points to the RFC.)